### PR TITLE
[CBRD-21804] Separate error context management from thread

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-21804

Separate error context management from thread.

Thread entry continues to store an error context, it registers its own context to error manager on start and deregisters on stop.

Error manager accesses registered context in thread local variable. Functions exposed by error manager that require context, like er_set, er_errid, er_stack_push/er_stack_pop will internally get thread local context and then manipulate it. Callers will continue to be oblivious of error context management.

This patch should also fix the memory leaks reported in issue.

A complete list of changes:
  1. Error context functionality moved to error_context.hpp/cpp
    * base level of er_message, er_message pointer to current error level.
    * context registration
    * stack internal management
    * static functions to get current context and current error level.
    * er_message structure was also moved here, some functions (clear error, reserve memory, so on) and message buffer was included.
   2. Error manager
    * fixed er_final not being called in some routines.
    * non SERVER_MODE will construct/destruct (and register/deregister) context during initialization/destruction.
    * moved er_start functionality that had nothing to do with context to er_init.
    * thread_get_thread_entry_info was replaced by cuberr::context::get_thread_local_context
    * manager class to be used for safe error manager initialization/final
  3. Other
    * master threads use their own error context (no longer dependent on thread entry to have your own error context)
    * logtb_find_client_tran_name_host_pid to get client transaction ID and other info.

This is not the final version of this patch. Some more change will follow (e.g. replacing ERROR LOG critical section with a mutex, code cleanup, so on).

However, I would like to start review and manual regression tests as soon as possible.

### CODE REVIEW

**Comment by ribram** [comment]
The medium regression seems to be actually correct now.

**Comment by ribram** [comment]
TODO: make error manager thread safe. Currently its thread safety is limited to SERVER_MODE only.

However, there are at least two other cases of error multi-threaded usage: master and broker.

**Comment by ribram** [comment]
Fixed shell crashes by:

  1. initializing error manager before using er_set/er_log functions in net_server_start, utility_initialize, boot_shutdown_client_at_exit, main of util_service
  2. register error context before using er_set/er_log functions in net_server_start, hb_thread_master_reader
  3. replace critical sections with standard mutexes.

@eseokoh 

If you think it helps, maybe we can run shell manually one more time.

**Comment by ribram** [comment]
There are two crashes left in shell_debug:

1. > _06_issues/_12_1h/bug_bts_6788/cases/bug_bts_6788.sh
  Fixed unitialized error manager in server_monitor_thr_f .

2. > _06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh
  Expected fatal error, because page buffer size parameter is too large. I am not sure why the issue is reported as NOK, but maybe QA can check it.

I think the PR is ready for review.

**Comment by eseokoh** [comment]
_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh

it expects an error is set for csql.err:
>    r2=`grep "214853222400" csql.err | wc -l`

It seems the case complains csql.err is empty.

**Comment by eseokoh** [comment]
please also notice a failure from medium test:
https://github.com/CUBRID/cubrid-testcases/blob/develop/medium/_05_err_x/cases/label1.sql

File: src/base/error_context.cpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    , msg_buffer {'\0'}
+    , m_logging (logging)
+  {
+    //
+  }
+
+  er_message::~er_message ()
+  {
+    ERMSG_LOG ("destruct ", *this);
+    clear_message_area ();
```
Comment (eseokoh) [inline src/base/error_context.cpp]: here we need to also free `args` and reset `nargs`.

Comment (ribram) [reply src/base/error_context.cpp]: done

**Pull request merged**